core/src/contracts/stablex_contract.rs

```diff
@@ -437,4 +516,49 @@ pub mod tests {
             (expected_prices, expected_token_ids)
         );
     }
+
+    #[test]
+    fn incremental_binary_search() {
+        struct MockBlockchain {
```

Comment:
Instead of implementing you own mock, you should be able to "automock" the trait like here:

https://github.com/gnosis/dex-services/blob/f1a0b703acb7512b767644415e12a2c6d1597b8e/core/src/contracts/stablex_contract.rs#L78-L79

and then configure the autogenerated mock with whatever you want to return like here:

https://github.com/gnosis/dex-services/blob/f1a0b703acb7512b767644415e12a2c6d1597b8e/core/src/solution_submission.rs#L263-L268

I see that here you are returning different values depending on the input. This should be also possible in automock as you have access tot he argument and can just look it up in a Mapping or list.